doc/api/n-api.md

```diff
@@ -0,0 +1,3150 @@
+
+#### *napi_create_string_utf8*
+
+##### Signature
+```C
+napi_status napi_create_string_utf8(napi_env env,
+                                    const char* str,
+                                    size_t length,
+                                    napi_value* result)
+```
+
+##### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  s`: Character buffer representing a UTF8-encoded string
+- `[in]  length`: The length of the string in characters, or -1 if it is
+null-terminated.
+- `[out] result`: A `napi_value` representing a JavaScript String
+
+##### Return value
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API creates a JavaScript String object from a UTF8-encoded C++ string
+
+The JavaScript String type is described in
+[Section 6.1.4](https://tc39.github.io/ecma262/#sec-ecmascript-language-types-string-type)
+of the ECMAScript Language Specification.
+
+### Functions to convert from N-API to C++ types
+
+#### *napi_get_array_length*
+
+##### Signature
+```C
+napi_status napi_get_array_length(napi_env env,
+                                  napi_value value,
+                                  uint32_t* result)
+```
+
+##### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  value`: `napi_value` representing the JavaScript Array whose length is
+being queried
+- `[out] result`: `uint32` representing length of the array
+
+##### Return value
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API returns the length of an array.
+
+Array length is described in
+[Section 22.1.4.1](https://tc39.github.io/ecma262/#sec-properties-of-array-instances-length) 
+of the ECMAScript Language Specification.
+
+#### *napi_get_arraybuffer_info*
+
+##### Signature
+```C
+napi_status napi_get_arraybuffer_info(napi_env env,
+                                      napi_value arraybuffer,
+                                      void** data,
+                                      size_t* byte_length)
+```
+
+##### Parameters
+- `[in]  env`: The environment that the API is invoked under
+- `[in]  arraybuffer`: `napi_value` representing the ArrayBuffer being queried
+- `[out] data`: The underlying data buffer of the ArrayBuffer
+- `[out] byte_length`: Length in bytes of the underlying data buffer
+
+##### Return value
+- `napi_ok` if the API succeeded.
+
+##### Description
+This API is used to retrieve the underlying data buffer of an ArrayBuffer and
+its length.
+WARNING: Use caution while using this API. The lifetime of the underlying data
+buffer is managed by the ArrayBuffer even after it's returned to you. A
+possible safe way to use this API is in conjunction with napi_create_reference,
+which can be used to guarantee that you control the lifetime of the
+ArrayBuffer. It's also safe to use the returned data buffer within the same
+callback as long as you don't call any other APIs that might trigger a GC.
+
+<TODO: What happens to this buffer in case of a compacting GC?>
```

Comment:
If you’re talking about the data pointer possibly being moved: That does [not](https://github.com/nodejs/node/blob/bde26a62bfeb70b642b04bed401c03ba6f1de936/src/node.cc#L4268-L4272) happen. :)